File: cpp/n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

// Recusive (DFS-ish)
class Solution {
private:
    void traverse_tree(Node* root, vector<int>& res) {
        if (!root)
            return ;

        res.push_back(root->val);

        for (auto child : root->children)
            traverse_tree(child, res);
    }

public:
    vector<int> preorder(Node* root) {
        vector<int> res;

        traverse_tree(root, res);

        return res;
    }
};

// Iteratively, with a stack
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> ret;
        if (!root) return ret;

        stack<Node*> s;
        s.push(root);
        while (!s.empty()) {
            Node* curr = s.top();
            s.pop();
            ret.push_back(curr->val);

            for (int i = curr->children.size()-1; i >=0; --i)
                if (curr->children[i])
                    s.push(curr->children[i]);
        }
        return ret;
    }
};


// Helper functions
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> values;
        recurse(root, values);
        iterate(root, values);
        return values;
    }

    void recurse(Node* node, vector<int>& values) {
        if (!node) return ;

        values.push_back(node->val);
        for (auto node : node->children) {
            recurse(node, values);
        }
    }

    void iterate(Node* root, vector<int>& values) {
        if (!root) return ;
    }
};

// Withe a deque, iteratively
class Solution {
public:
    vector<int> preorder(Node* root) {
        if (!root) return {};
        vector<int> values;
        iterate(root, values);
        return values;
    }

    void iterate(Node* root, vector<int>& values) {
        if (!root) return ;

        deque<Node*> nodes;
        nodes.push_front(root);

        while (!nodes.empty()) {
            const auto node = nodes.front();
            nodes.pop_front();
            values.push_back(node->val);

            // Traverse with reverse iterators
            for (auto itr = node->children.rbegin(); itr != node->children.rend(); ++itr) {
                nodes.push_front(*itr);
            }
        }
    }
};



class Solution {
public:
  vector<int> preorder(Node* root) {
    vector<int> values;
    if (root) preorder(root, values);
    return values;
  }
    
  void preorder(Node* node, vector<int>& values) {
    values.push_back(node->val);
    for (const auto& node : node->children) {
      preorder(node, values);
    }
  }
};

File: py/n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
# basic recursive solution
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        res = [root.val]
        for child in root.children:
            result.extend(self.preorder(child))
        return result

# inefficient space solution since we store every node, but much faster
# than most approaches to this problem
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []

        nodes = [root]
        out = []

        while nodes:
            temp = nodes.pop()
            out.append(temp.val)

            for child in reversed(temp.children):
                nodes.append(child)
        return out

# interesting assignment to 'nodes' makes this code look much cleaner/pythonic
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        ret = []
        nodes = root and [root]
        while nodes:
            node = nodes.pop()
            ret.append(node.val)
            nodes += [child for child in reversed(node.children) if child]
        return ret

# Using a stack
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        stack = [root]
        ret = []
        while stack:
            node = stack.pop()
            if node:
                ret.append(node.val)
                stack += [child for child in reversed(node.children) if child]
        return ret

# Helper function (DFS-ish)
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        ret = []
        self.helper(root, ret)
        return ret
    def helper(self, root, ret: 'Node') -> None:
        if not root:
            return
        ret.append(root.val)
        if root.children:
            for kid in root.children:
                self.helper(kid, ret)
            # map(self.helper, root.children, ret) # didn't work
        return

# DFS, only pass the node by defining the function inside preorder
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        def dfs(node: 'Node') -> None:
            if not node:
                return
            ans.append(node.val)
            for c in node.children:
                dfs(c)
        ans = []
        dfs(root)
        return ans










